### DISCUSSION TITLE
Unable to build Theia extension which uses localization (Can't resolve 'fs' in '...\node_modules\jsonfile')

### DISCUSSION THREAD

**Discussion by BarryDrez** [eclipse-theia/theia#10637] [General]
### Bug Description:
I am unable to build my Theia extension which uses localization.  I am seeing the error: Can't resolve 'fs' in '...\node_modules\jsonfile'

### Steps to Reproduce:

1. Followed the instructions listed in Theia documentation under [Internationalization](https://theia-ide.org/docs/i18n/)
2. Build my extension using yarn
3. Build fails with error:
> ERROR in ./node_modules/jsonfile/index.js 5:8-21
Module not found: Error: Can't resolve 'fs' in 'C:\Development\theia\java-service-editor\node_modules\jsonfile'

### Additional Information

- Operating System:  Windows 10
- Theia Version: 1.21.0
- Node Version: 17.3.1
- yarn Version 1.22.5
- webpack Version 5.66.0
- webpack-cli Version 4.7.0

**Comment by vince-fugnitto** [comment]
@BarryDrez thank you for the discussion, I'm not sure it is related to `localization` at first glance, are you using `fs` (node apis) in the browser (frontend)?

**Reply by BarryDrez** [reply]
Hi @vince-fugnitto.  I am not using fs anywhere.  The error occurs as soon as I import LocalizationContribution from '@theia/core/lib/node/i18n/localization-contribution' and bind it in my frontend module.  Thank you.

**Reply by BarryDrez** [reply]
I should have provided the complete error:
ERROR in ./node_modules/jsonfile/index.js 5:8-21
Module not found: Error: Can't resolve 'fs' in 'C:\Development\theia\my-extension\node_modules\jsonfile'
 @ ./node_modules/@theia/core/node_modules/fs-extra/lib/json/jsonfile.js 4:17-36
 @ ./node_modules/@theia/core/node_modules/fs-extra/lib/json/index.js 4:17-38
 @ ./node_modules/@theia/core/node_modules/fs-extra/lib/index.js 14:11-28
 @ ./node_modules/@theia/core/lib/node/i18n/localization-contribution.js 28:11-30
 @ ./node_modules/java-service/lib/browser/java-service-frontend-module.js 18:36-98
 @ ./src-gen/frontend/index.js 75:31-94

jsonfile is in Theia dependencies.

**Reply by BarryDrez** [reply]
To follow up, I added the following to my webpack.config.js:
`config.resolve.fallback = {
    fs: false,
};`
and that problem went away, but 9 other similar errors appeared.  So I added those to my webpack.config.js which now looks like:
`config.resolve.fallback = {
    fs: false,
    os: false,
    net: false,
    path: false,
    util: false,
    assert: false,
    crypto: false,
    stream: false,
    constants: false,
    child_process: false
};`
and now I can build.  However, when I run my application, it does not start, and I see the following error in the console:

> react_devtools_backend.js:4045 ReferenceError: process is not defined
    at Object../node_modules/graceful-fs/polyfills.js (polyfills.js:3:1)
    at __webpack_require__ (bootstrap:19:1)
    at Object../node_modules/graceful-fs/graceful-fs.js (graceful-fs.js:2:17)
    at __webpack_require__ (bootstrap:19:1)
    at Object../node_modules/@theia/core/node_modules/fs-extra/lib/fs/index.js (index.js:4:12)
    at __webpack_require__ (bootstrap:19:1)
    at Object../node_modules/@theia/core/node_modules/fs-extra/lib/index.js (index.js:8:12)
    at __webpack_require__ (bootstrap:19:1)
    at Object../node_modules/@theia/core/lib/node/i18n/localization-contribution.js (localization-contribution.ts:17:1)
    at __webpack_require__ (bootstrap:19:1)

**Reply by vince-fugnitto** [reply]
@BarryDrez 

> The error occurs as soon as I import LocalizationContribution from '@theia/core/lib/node/i18n/localization-contribution' and bind it in my frontend module. 

That should be the issue, you’re attempting to import `node` (backend) code in the browser/frontend.

**Reply by BarryDrez** [reply]
@vince-fugnitto 
The documentation specifies: 

> Lastly, this LocalizationContribution will have to be bound within your **frontend** injection module:
> bind(CustomLocalizationContribution).toSelf().inSingletonScope();
bind(LocalizationContribution).toService(CustomLocalizationContribution);

is this correct?

**Reply by vince-fugnitto** [reply]
It's possibly a documentation error, @msujew can you confirm?

**Reply by BarryDrez** [reply]
I moved my LocalizationContribution implementation and the bindings to the backend module, and it builds, and my application now runs in the browser; however my keys are not getting translated.
While debugging theia/core/src/common/nls.ts, I can see that the localization variable is undefined.

**Reply by BarryDrez** [reply]
I ran the command **Configure Display Language** and selected en.  Now, the localization variable is there.  The problem now is that it is not picking up my resources.  I have defined them in a file: nls.en.json in a folder data/i18n.  My LocalizationContribution implementation has the line:
`registry.registerLocalizationFromRequire('en', require('../../data/i18n/nls.en.json'));`
I don't think it is being loaded.

**Reply by msujew** [reply]
@vince-fugnitto Yep, definitely a documentation error!

@BarryDrez You don't need a localization file for English, only for other languages. The English value will be picked up automatically from the default value provided to the `nls.localize(key, defaultValue, args)` call. 

Just an FYI from a technical perspective: The feature is written in such a way that for the default case (English language selected), no localizations are retrieved from the server, instead only the default values (which should be in English) are used in the frontend.

**Reply by BarryDrez** [reply]
Thanks @vince-fugnitto Is that the only way to handle English?  We have all of our strings (including English) in a database and a localization team that translates into many different languages.  Our resource files (such as nls.en.json) are generated from the database.  If we have to use the default values, that means we must maintain the strings in 2 different places.  Is it not possible to use localizeByDefault?  Is there any way to load the English bundle so I can only pass a key?

**Reply by BarryDrez** [reply]
@vince-fugnitto  I have added a German resource file (nls.de.json), and changed my display language to de, and now my English text from nls.en.json is being displayed.  While debugging, I can see that the languageId is still "en".

**Reply by msujew** [reply]
@BarryDrez Sure, you can put your own translation function in front of `nls.localize`, which tries to first resolve to your `nls.en.json` file. It could look like this:

```
import { nls } from '@theia/core';
import { Localization, FormatType } from '@theia/core/lib/common/i18n/localization';

const enValues = require('../../data/i18n/nls.end.json');

export function nlsWithEnglish(key: string, ...args: FormatType[]): string {
  if (nls.locale === undefined || nls.locale === 'en') {
    return Localization.format(enValues[key], args);
  }
  return nls.localize(key, '', ...args);
}
```

Be aware that the `theia nls-extract` CLI command will cease to function, since it depends on static analysis of your code.

For the other issue, I can't reproduce this. After installing the german language pack and configuring the display language, my `localeId` in the local storage is `de`:

![image](https://user-images.githubusercontent.com/4377073/149514738-fc78ae41-9d61-456f-bda9-e1f908688590.png)

And the `nls.locale` constant is also correctly set:

![image](https://user-images.githubusercontent.com/4377073/149514880-62e0bb4c-72df-4e4f-8d49-db0617797011.png)

**Reply by BarryDrez** [reply]
@msujew Thank you for the detailed explanation.  Actually, I am calling nls.localize('my.resource.key', ''), and it is getting my English string from nls.en.json correctly.  This started as when I added the German translation file.

With regards to displaying German text, I have added the following:
```
	async registerLocalizations(registry: LocalizationRegistry): Promise<void> {
		registry.registerLocalizationFromRequire('en', require('../../data/i18n/nls.en.json'));
		registry.registerLocalizationFromRequire('de', require('../../data/i18n/nls.de.json'));
	}
```
to my LocalizationContribution implementation, and I added a nls.de.json file in the same location as my nls.en.json file.  However, when I run the Configure Display Language command, only en appears in the list of options to select from.  Could there be a step that I missed?

**Reply by msujew** [reply]
The `Configure Display Language` command is coded in such a way, that only languages can be selected which are installed using a VSCode language pack. This is done to prevent the application from being only partly translated into German/other languages, since Theia's internal translations are derived from VSCode's language packs. You can download the German language pack using the vsx-registry view or the [open-vsx marketplace directly](https://open-vsx.org/extension/MS-CEINTL/vscode-language-pack-de).

After installing that one, you should be able to select German as a display language.

**Reply by BarryDrez** [reply]
@msujew I see.  That should be fine then.  Thanks very much to you and @vince-fugnitto for all your help!

**Reply by msujew** [reply]
Great :) You're welcome!
